src/common/transformations/src/transformations/sdpa_to_paged_attention/state_management_pattern.cpp

```diff
@@ -149,16 +154,59 @@ ov::pass::StateManagementPattern::StateManagementPattern(ParameterVector& kv_par
 
         auto real_k = take_4d(k_current, k_current_reshaped, k_current2);
         auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+        auto sdpa_node = pattern_map.at(sdpa).get_node();
+        // E and Ev are from the SDPA specification at
+        // https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/operation-specs/sequence/scaled-dot-product-attention.html
+        auto E = sdpa_node->get_input_tensor(1).get_partial_shape()[-1];
+        auto Ev = sdpa_node->get_input_tensor(2).get_partial_shape()[-1]; // in common case may not match E
+        auto num_q_heads = sdpa_node->get_input_tensor(0).get_partial_shape()[-3];
+
+        auto extract_num_kv_heads = [=, &pattern_map](std::shared_ptr<Node>& unsqueeze) {
+            // Deduce number of k/v heads from Unsqueeze-Broadcast-Reshape (if present) pattern that appears in case of MQA/GQA
+            if (pattern_map.find(unsqueeze) != pattern_map.end()) {
+                // based on unsqueeze index determine the dimension that will be broadcased
+                // if there is no expected dimension for any reason, return dynamic dimension
+                unsqueeze = pattern_map.at(unsqueeze).get_node_shared_ptr(); //we reassign the pointer. Ok?
+                auto shape = unsqueeze->get_output_partial_shape(0);
+                auto rank = shape.rank();
+                if (rank.is_dynamic()) {
+                    return ov::Dimension();
+                }
+                rank = rank.get_length();
+                auto axis = unsqueeze->input_value(1).get_node_shared_ptr();
+                auto constant = std::dynamic_pointer_cast<ov::op::v0::Constant>(axis);
+                if (!constant) {
+                    return ov::Dimension();
+                }
+                auto data = constant->get_vector<int64_t>(); //hope it's correct here ??
+                if (data.size() != 1) { // it should be only one axis
+                    return ov::Dimension();
+                }
+                auto first_element = data[0];
+                if (first_element == 0 || first_element == -rank.get_length()) { // there should be at least one dimension to the left
+                    // -rank.get_length() correct??
```

Comment:
Is `-rank.get_length()` what meant in the Python code?

Reply:
yes